File: src/test/java/no/nav/tjenestepensjon/simulering/consumer/TokenClientTest.java
package no.nav.tjenestepensjon.simulering.consumer;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.client.WireMock;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import no.nav.tjenestepensjon.simulering.domain.Token;

class TokenClientTest {
    private static TokenClient tokenClient = new TokenClient();

    private static WireMockServer wireMockServer;

    @BeforeAll
    static void beforeAll() {
        wireMockServer = new WireMockServer();
        wireMockServer.start();
        tokenClient.setUsername("username");
        tokenClient.setPassword("password");
        tokenClient.setStsUrl("http://localhost:8080");
    }

    @Test
    void shouldGetSamlToken() {
        wireMockServer.stubFor(WireMock.get(WireMock.urlPathEqualTo("/rest/v1/sts/samltoken"))
                .willReturn(WireMock.okJson("{\"access_token\":\"eyJ4vaea3\",\"expires_in\":\"3600\",\"token_type\":\"Bearer\",\"issued_token_type\":\"saml:blah:2.0\"}")));

        Token token = tokenClient.getSamlAccessToken();

        assertThat(token.getAccessToken(), is("eyJ4vaea3"));
        assertThat(token.getExpiresIn(), is(3600L));
        assertThat(token.getTokenType(), is("Bearer"));
        assertThat(token.getIssuedTokenType(), is("saml:blah:2.0"));
    }

    @Test
    void shouldGetOidcToken() {
        wireMockServer.stubFor(WireMock.get(WireMock.urlPathEqualTo("/rest/v1/sts/token"))
                .willReturn(WireMock.okJson("{\"access_token\":\"eyJ4vaea3\",\"expires_in\":\"3600\",\"token_type\":\"Bearer\"}")));

        Token token = tokenClient.getOidcAccessToken();

        assertThat(token.getAccessToken(), is("eyJ4vaea3"));
        assertThat(token.getExpiresIn(), is(3600L));
        assertThat(token.getTokenType(), is("Bearer"));
        assertThat(token.getIssuedTokenType(), is(nullValue()));
    }

    @Test
    void shouldFailWhenStatusNot200() {
        wireMockServer.stubFor(WireMock.get(WireMock.urlPathEqualTo("/rest/v1/sts/token"))
                .willReturn(WireMock.badRequest().withBody("Bad request!")));

        assertThrows(RuntimeException.class, () -> tokenClient.getOidcAccessToken());
    }

    @AfterAll
    static void afterAll() {
        wireMockServer.stop();
    }
}

File: src/test/kotlin/no/nav/tjenestepensjon/simulering/consumer/TokenClientTest.kt
package no.nav.tjenestepensjon.simulering.consumer

import AsyncExecutor.AsyncResponse
import OutgoingResponse.SimulertPensjon
import com.github.tomakehurst.wiremock.WireMockServer
import com.github.tomakehurst.wiremock.client.WireMock
import no.nav.ekstern.pensjon.tjenester.tjenestepensjonsimulering.meldinger.v1.Stillingsprosent
import no.nav.tjenestepensjon.simulering.domain.Token
import org.hamcrest.Matchers
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

internal class TokenClientTest {
    @Test
    fun shouldGetSamlToken() {
        wireMockServer!!.stubFor(WireMock.get(WireMock.urlPathEqualTo("/rest/v1/sts/samltoken"))
                .willReturn(WireMock.okJson("{\"access_token\":\"eyJ4vaea3\",\"expires_in\":\"3600\",\"token_type\":\"Bearer\",\"issued_token_type\":\"saml:blah:2.0\"}")))
        val token: Token = tokenClient.getSamlAccessToken()
        assertThat(token.getAccessToken(), Matchers.`is`("eyJ4vaea3"))
        assertThat(token.getExpiresIn(), Matchers.`is`(3600L))
        assertThat(token.getTokenType(), Matchers.`is`("Bearer"))
        assertThat(token.getIssuedTokenType(), Matchers.`is`("saml:blah:2.0"))
    }

    @Test
    fun shouldGetOidcToken() {
        wireMockServer!!.stubFor(WireMock.get(WireMock.urlPathEqualTo("/rest/v1/sts/token"))
                .willReturn(WireMock.okJson("{\"access_token\":\"eyJ4vaea3\",\"expires_in\":\"3600\",\"token_type\":\"Bearer\"}")))
        val token: Token = tokenClient.getOidcAccessToken()
        assertThat(token.getAccessToken(), Matchers.`is`("eyJ4vaea3"))
        assertThat(token.getExpiresIn(), Matchers.`is`(3600L))
        assertThat(token.getTokenType(), Matchers.`is`("Bearer"))
        assertThat(token.getIssuedTokenType(), Matchers.`is`(Matchers.nullValue()))
    }

    @Test
    fun shouldFailWhenStatusNot200() {
        wireMockServer!!.stubFor(WireMock.get(WireMock.urlPathEqualTo("/rest/v1/sts/token"))
                .willReturn(WireMock.badRequest().withBody("Bad request!")))
        Assertions.assertThrows(RuntimeException::class.java) { tokenClient.getOidcAccessToken() }
    }

    companion object {
        private val tokenClient: TokenClient = TokenClient()
        private var wireMockServer: WireMockServer? = null
        @BeforeAll
        fun beforeAll() {
            wireMockServer = WireMockServer()
            wireMockServer!!.start()
            tokenClient.setUsername("username")
            tokenClient.setPassword("password")
            tokenClient.setStsUrl("http://localhost:8080")
        }

        @AfterAll
        fun afterAll() {
            wireMockServer!!.stop()
        }
    }
}